log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableMessageFactory.java

```diff
@@ -30,6 +30,7 @@
  * @since 2.6
  */
 @PerformanceSensitive("allocation")
+@SuppressWarnings("ThreadLocalUsage")
```

Comment:
Heh, this should end up getting removed once we finalize the recycler API!

Reply:
Sorry I didn't add the meaning of all these warnings in comments (cf. [Error Prone documentation](http://errorprone.info/bugpatterns))

[ThreadLocalUsage](http://errorprone.info/bugpattern/ThreadLocalUsage) actually means: only use `static` thread locals.

I believe we should use instance thread locals for singleton's like this. We only need to ensure that this is a singleton (except in tests). E.g. if a user asks for the message factory `org.apache.logging.log4j.message.ReusableMessageFactory`, we can do like `ServiceLoader` in Java 9+: look for a static `#provider` method and only if this is not found instantiate the class.

Reply:
Ugh, that rule is more of a warning than an error as it's easy to accidentally use up far more memory than expected depending on the usage pattern. Using a static thread local field can also leak memory.

Reply:
I agree, all of the changes in this PR are warnings and some code changes are there just to help Error Prone understanding that we understand the problem.